parlai/core/torch_generator_agent.py

```diff
@@ -875,15 +877,26 @@ def eval_step(self, batch):
             beam_preds_scores, beams = self._generate(
                 batch, self.beam_size, maxlen, prefix_tokens=prefix_tokens
             )
-            preds, scores = zip(*beam_preds_scores)
+            preds, _, _, _= zip(*beam_preds_scores)
             self._add_generation_metrics(batch, preds)
 
             # bsz x beamsize
             beam_texts: List[List[Tuple[str, float]]] = []
+            beam_generated_text_token_info: List[List[Tuple[str, float, int]]] = []
+
             for beam in beams:
                 beam_texts.append([])
-                for tokens, score in beam.get_rescored_finished():
+                if self.output_token_losses:
+                    beam_generated_text_token_info.append([])
+
+                for tokens, score, token_scores, token_ranks in beam.get_rescored_finished():
                     try:
+                        token_scores_pretty = None
+                        if self.output_token_losses:
+                            tokens_as_txt = [self.dict[int(token)] for token in tokens]
+                            token_scores_pretty = list(zip(tokens_as_txt, token_scores.tolist(), token_ranks.tolist()))
+                            beam_generated_text_token_info[-1].append(token_scores_pretty)
```

Comment:
I'm nervous that we're doing some slow per-item gpu copies here. Can you run a benchmark of `eval_model` on some dataset before and after this change?

Reply:
Will do. Will something like `time parlai eval_model --task "babi:Task1k:2" -m "repeat_label"` to check how long `eval_model` took be okay, or did you have something else in mind?

Reply:
Not sure 

On my code:
```
> time parlai eval_model --task "babi:Task1k:2" -m "repeat_label"
...
real    0m3.068s
user    0m3.349s
sys     0m5.302s
```
On `main`:
```
> time parlai eval_model --task "babi:Task1k:2" -m "repeat_label"
...
real    0m3.394s
user    0m3.341s
sys     0m5.162s
```

Reply:
This must be measuring noise, because `-m repeat_label` isn't hitting any of your code paths.

You need to test with an actual generative model.

Reply:
Updated! I test with a transformer model now and turn on verbose mode with `-v` to ensure the generative code is run

Reply:
It looks like you were right. There seems to be a nontrivial degradation in performance. Any tips for how to improve this? What kinds of difference in performance would you be okay with?

Reply:
Also, my code updates only run if `verbose` mode is on. If you really care about performance, you would probably turn this mode off anyways.

Reply:
Curious -- we're not copying any big param matrices. We're just copying utterances that will probably be pretty short (the number of beams will be pretty small too). What made you think this could be an issue?

Reply:
@emilydinan thoughts?